### PR TITLE
Initial commit for ML ops guide

### PR DESCRIPTION
**Pull request by odette-elementl**
## Summary & Motivation

> **Does this pull request update the docs?** If so, please verify that the updates follow the [docs style checklist](https://github.com/dagster-io/dagster/blob/master/docs/DOC_CHECKLIST.md).

link to the page https://dagster-git-odette-mlops-elementl.vercel.app/guides/dagster/ml-pipeline

---

### CODE REVIEW

**Review by sryza — COMMENTED** [review]
Excited about this! I did an initial pass over the python file - I haven't yet looked at the MDX file.

What do you think about putting this into the examples directory? I bet that some users would benefit from being able to copy it as a full project and modify it from there.

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py
[the remark this answers is not in the document]
Comment (odette-elementl) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:1]: ok will add

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+@asset
+def transform_train(test_train_split):
+    X_train, X_test, y_train, y_test = test_train_split
+    vectorizer = TfidfVectorizer()
+    transformed_X_train = vectorizer.fit_transform(X_train)
+    transformed_X_train = transformed_X_train.toarray()
+    y_train = y_train.fillna(0)
+    transformed_y_train = np.array(y_train)
+
+    return vectorizer, transformed_X_train, transformed_y_train
```
[the remark this answers is not in the document]
Comment (odette-elementl) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:82]: thought would be that if you wanted to take a new dataset and transform it to do predictions you would want to use it

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    Output,
+    FreshnessPolicy,
+)
+import datetime
+from sklearn.model_selection import train_test_split
+from sklearn.feature_extraction.text import CountVectorizer
+from sklearn import linear_model
+
+
+@asset
```
[the remark this answers is not in the document]
Comment (odette-elementl) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:19]: yeah that works was trying to see how to make it more dynamic to load and add incrementally but i think for this case its too much ETL for an ML guide

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+## test_train_split_end
+
+
+## vectorizer_start
+@asset
+def transform_train(test_train_split):
+    X_train, X_test, y_train, y_test = test_train_split
+    vectorizer = TfidfVectorizer()
+    transformed_X_train = vectorizer.fit_transform(X_train)
```
[the remark this answers is not in the document]
Comment (sryza) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:77]: ah, makes sense

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+## vectorizer_end
+
+## models_start
+
+from sklearn.tree import DecisionTreeRegressor
+
+
+@asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=1440))
+def decision_tree(transform_train, transform_test):
```
[the remark this answers is not in the document]
Comment (sryza) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:104]: my instinct would be to go with xgboost, because:
- it's state of the art
- even though it's a more complicated model internally, it doesn't actually make the code more complex

[the remark this answers is not in the document]
Comment (odette-elementl) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:104]: sounds good

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sryza) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:82]: Ah, I see. It seems like we basically need it in any place we want to do inference? If that's the case, what do you think about combining it with the predictive model itself into a class called something like `StoryTitleCommentsPredictor`? Then, anywhere we need to do inference, we can just depend on that single asset.

Also, I know it adds some complexity, but I think it could be worth adding an inference step to the pipeline? Maybe in a separate Python file?

**Review by erinkcochran87 — COMMENTED** [review]
Digging where this is going, Odette! Some general thoughts, in no particular order:

- There should be a **Prerequisites** section, as it seems users are expected to have familiarity with ML, not just Dagster.
- To build on that ^, should users have completed the tutorial before starting this guide?
- Would it be useful to define some ML terms before diving into the guide? Ex: `tokenize`, `validation set`, etc.
- Would showing Dagit in here be useful at all?
- We try to include imports in code examples whenever possible to make them as copy-and-pasteable as possible

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (odette-elementl) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py]: added inference asset - do we want it to predict or score a new dataset?

File: docs/content/guides/dagster/ml-ops.mdx [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+        df = df[df.type == "story"]
+        df = df[~df.title.isna()]
+    inference_x = df.title
+    inference_x = Tfidf_Vectorizer.transform(inference_x)
+    return xgboost.predict(inference_x)
+```
+
+Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history and trigger retraining.
+
+Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.
```
Comment (odette-elementl) [inline docs/content/guides/dagster/ml-ops.mdx:196]: was between scoring and predicting here..

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+## vectorizer_end
+
+
+## models_start
+
+import xgboost as xg
+from sklearn.metrics import mean_absolute_error
+
+@asset
+def xgboost(transformed_training_data):
```
Comment (sryza) [inline examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:83]: In general, assets should be named in terms of the object they represent - i.e. what would be the name of the file if you wanted people who come across it to know what it is?  Something like:
- `comments_model`
- `xgboost_comments_model`
- `xgboost_num_comments_regressor`

Depending on how verbose we want to be.

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    transformed_X_test, transformed_y_test = transformed_test_data
+    """Use the test set data to get a score of the XGBoost model"""
+    score = xgboost.score(transformed_X_test, transformed_y_test)
+    return score 
+## models_end
+
+
+## inference_start
+@asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=60))
+def model_inference(xgboost, Tfidf_Vectorizer):
```
Comment (sryza) [inline examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:102]: Same comment as above on this one: In general, assets should be named in terms of the object they represent - i.e. what would be the name of the file if you wanted people who come across it to know what it is?

Maybe something like `latest_story_comment_predictions`?

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+def xgboost(transformed_training_data):
+    transformed_X_train, transformed_y_train = transformed_training_data
+    """Train XGBoost model, which is a highly efficent and flexible model"""
+    xgb_r = xg.XGBRegressor(objective ='reg:squarederror', eval_metric=mean_absolute_error,
+                  n_estimators = 20)
+    xgb_r.fit(transformed_X_train, transformed_y_train)
+    return xgb_r
+
+@asset 
+def score_xgboost( transformed_test_data, xgboost):
```
Comment (sryza) [inline examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:92]: Same comment as above on this one: In general, assets should be named in terms of the object they represent - i.e. what would be the name of the file if you wanted people who come across it to know what it is?

Maybe something like `comments_model_test_set_r_squared`? Or `comments_model_test_set_evaluation` if we want to include other evaluation metrics too, like MSE.

**Review by sryza — COMMENTED** [review]
I haven't had a chance to look at the latest version of the guide yet, but I left a few more comments on the example.  It looks close!  I plan to get to the guide next week.

I think parts of these earlier comments are still applicable:
- https://github.com/dagster-io/dagster/pull/12930#discussion_r1134703805
- https://github.com/dagster-io/dagster/pull/12930#discussion_r1137284340

File: examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
+
+## data_ingestion_start
+import requests
+from dagster import asset, FreshnessPolicy
+import pandas as pd
+
+@asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=10))
+def hackernews_stories():
+    """Get the max ID number from hacker news"""
```
[the remark this answers is not in the document]
Comment (odette-elementl) [reply examples/docs_snippets/docs_snippets/guides/dagster/ml_ops/ml_ops.py:9]: sounds good will change wasnt sure if that was a style thing because i saw in the tutorial

File: docs/content/guides/dagster/ml-ops.mdx [hunk 10]
```diff
@@ -0,0 +1,221 @@
+---
+title: Building your machine learning pipelines In Dagster | Dagster Docs
+description: This guide illustrates how to use Dagster to operationaize your machine learning pipeline
+---
+
+# Building your machine learning pipelines with Dagster
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:6]: I would take out "your"

File: docs/content/guides/dagster/ml-ops.mdx [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+
+In this guide, we’ll walk you through how to take your machine learning models and deploy and mantain them in production using Dagster, reliably and efficently.
+
+<Note>
+  This guide assumes you have familiarity with machine learning concepts,{" "}
+  <a href="/tutorial">the Dagster tutorial</a> and several Dagster concepts,
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:12]: I don't think we necessarily assume they have familiarity with the tutorial, just with the concepts. We can suggest that they take the tutorial if they're not familiar with the concepts.

File: docs/content/guides/dagster/ml-ops.mdx [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+## Why build your ML pipelines with Dagster?
+
+- Dagster is designed to use during development, the lightweight execution model means you can access the benefits of an orchestrator, like re-executing from the middle of a pipeline and parallelizing steps while you're experiementing.
+- It models data assets, not just tasks so Dagster understands the upstream and downstream data depenedencies and is a one-stop shop for both the data transformations and the models that depend on the data transformations.
+
+---
+
+## Machine learning development
+
+Before we go into how to operatialize a machine learning model, let's go over some of the tools you can use in Dagster to develop a machine learning model.
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:36]: What tools do you mean?

File: docs/content/guides/dagster/ml-ops.mdx [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+</Note>
+
+We will work through several aspects of MLOps, including using assets for different elements of your machine learning pipeline, how to automate model training and monitoring your model's drift.
+
+---
+
+## Why build your ML pipelines with Dagster?
+
+- Dagster is designed to use during development, the lightweight execution model means you can access the benefits of an orchestrator, like re-executing from the middle of a pipeline and parallelizing steps while you're experiementing.
+- It models data assets, not just tasks so Dagster understands the upstream and downstream data depenedencies and is a one-stop shop for both the data transformations and the models that depend on the data transformations.
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:30]:
```suggestion
- Dagster models data assets, not just tasks, so it understands the upstream and downstream data dependencies and is a one-stop shop for both the data transformations and the models that depend on the data transformations.
```

Comment (sryza) [reply docs/content/guides/dagster/ml-ops.mdx:30]: This is a bit of a run-on sentence.

File: docs/content/guides/dagster/ml-ops.mdx [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+  <a href="/concepts/assets/software-defined-assets">software-defined assets</a>{" "}
+  and <a href="/concepts/ops-jobs-graphs/jobs">jobs</a> . We also recommend
+  reviewing the{" "}
+  <a href="/guides/dagster/automating-pipelines">
+    automating data pipelines guide
+  </a>{" "}
+  as many of the same principles can be applied to MLOps.
+</Note>
+
+We will work through several aspects of MLOps, including using assets for different elements of your machine learning pipeline, how to automate model training and monitoring your model's drift.
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:23]: Oxford comma plz

File: docs/content/guides/dagster/ml-ops.mdx [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+<Image
+alt="alt"
+src="/images/guides/ml-ops/ml_asset_dag.png"
+width={1804}
+height={1064}
+/>
+
+### Data Ingestion
+
+First, we will be creating an asset which retrieves all the hackernews records from the previous run to the latest hackernews record.
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:55]: This isn't exactly accurate, correct? In the sense that we're not retrieving all the records from the previous run?

File: docs/content/guides/dagster/ml-ops.mdx [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+    return xgboost.predict(inference_x)
+```
+
+Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history and trigger retraining.
+
+Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.
+
+---
+
+## Machine learning operations
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:200]: My instinct would be to leave this out for the initial PR.  I think there's value in this kind of content, and I think this is a solid start, but I also think it will take some back and forth to get it right, and I'd rather not have this block the rest of the guide from making it in.

**Review by sryza — COMMENTED** [review]
I just did a pass on the guide section. I think the overall structure and most of the content looks great.

A few high-level comments:
- I recommend doing a self-review to fix spelling and grammar.
- The guide switches between first and second person. We should stick to one. @erinkcochran87 - did we decide on which we want to use for this kind of content?
- As I mentioned within a comment, I think we should leave the "Machine learning operations" section out of this PR.  I think there's high value in this kind of content, and I think this is a solid start, but I also think it will take some back and forth to get it right, and I'd rather not have this block the rest of the guide from making it in.

File: docs/content/guides/dagster/ml-ops.mdx (hunk 16 above)
Comment (odette-elementl) [reply docs/content/guides/dagster/ml-ops.mdx:200]: sounds good saving it in notion for later

File: docs/content/guides/dagster/ml-ops.mdx (hunk 15 above)
Comment (odette-elementl) [reply docs/content/guides/dagster/ml-ops.mdx:55]: updated

File: docs/content/guides/dagster/ml-ops.mdx (hunk 11 above)
Comment (erinkcochran87) [reply docs/content/guides/dagster/ml-ops.mdx:12]: As this guide builds on the tutorial, that's why there's familiarity suggested here. I agree we can rephrase though. Perhaps something along these lines?:

> This guide assumes you have familiarity with machine learning and several Dagster concepts. It also builds on the [asset tutorial](), which we recommend completing if you're unfamiliar with Dagster or want a comprehensive look at the project code.

File: docs/content/guides/dagster/ml-ops.mdx (hunk 10 above)
Comment (erinkcochran87) [reply docs/content/guides/dagster/ml-ops.mdx:6]: +1

File: docs/content/guides/dagster/ml-ops.mdx (hunk 11 above)
Comment (sryza) [reply docs/content/guides/dagster/ml-ops.mdx:12]: +1

File: docs/content/guides/dagster/ml-ops.mdx [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+        df = df[~df.title.isna()]
+
+    return df
+```
+
+### Data Transformation
+
+Now that we have a dataframe with all valid stories, we want to transform that data into something our machine learning model will be able to use.
+
+The first step is taking the dataframe and splitting into a [training and test set](https://en.wikipedia.org/wiki/Training,\_validation,\_and_test_data_sets). In some of your models, you also might choose to have an additional split for a validation set. The reason we split the data is so that we can have a test and/or a validation dataset that is independent of the training set and we can use that to see how well our model did.
```
Comment (erinkcochran87) [inline docs/content/guides/dagster/ml-ops.mdx:86]:
```suggestion
The first step is taking the dataframe and splitting into a [training and test set](https://en.wikipedia.org/wiki/Training,\_validation,\_and_test_data_sets). In some of your models, you also might choose to have an additional split for a validation set. The reason we split the data is so that we can have a test and/or a validation dataset that is independent of the training set. We can then use that dataset to see how well our model did.
```

🔧 Suggesting you break up the last sentence in this paragraph. It currently reads a bit like a run on and can be difficult to follow.

File: docs/content/guides/dagster/ml-ops.mdx [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+    return transformed_X_test, transformed_y_test
+```
+
+We have also transformed the dataframes into numpy arrays and removed na values to prepare the data for training.
+
+### Model training
+
+At this point, we have X_train, y_train, X_test, y_test ready to go for our model. We can use any number of models to train our model from libraries like [sklearn](https://scikit-learn.org/), [tensorflow](https://www.tensorflow.org/) and [pytorch](https://pytorch.org/).
+
+In our example, we will be training a [XGBoost model](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor) to predict a numerical value.
```
Comment (erinkcochran87) [inline docs/content/guides/dagster/ml-ops.mdx:136]:
```suggestion
In our example, we will train an [XGBoost model](https://xgboost.readthedocs.io/en/stable/python/python_api.html#xgboost.XGBRegressor) to predict a numerical value.
```

🔧 Suggesting we use `we will train` instead of `we will be training` - it's less wordy and more action-forward.

File: docs/content/guides/dagster/ml-ops.mdx [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+
+@asset
+def score_xgboost( transformed_test_data, xgboost):
+    transformed_X_test, transformed_y_test = transformed_test_data
+    """Use the test set data to get a score of the XGBoost model"""
+    score = xgboost.score(transformed_X_test, transformed_y_test)
+    return score
+```
+
+We include freshness policies in the models here which will ensure the upstream data transformation assets are updated to comply with these policies.
```
Comment (erinkcochran87) [inline docs/content/guides/dagster/ml-ops.mdx:159]: ❓ Include a link to freshness policy documentation?

File: docs/content/guides/dagster/ml-ops.mdx [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+
+### Evaluate our results
+
+In our model assets, we evaluated each of the models on the test data and in this case, got the error between the predicted and actual results. Next, lets create an asset which runs inference on the model, more frequently than the model is re-trained to predict the results.
```
Comment (erinkcochran87) [inline docs/content/guides/dagster/ml-ops.mdx:163]: > In our model assets, we evaluated each of the models on the test data and in this case, got the error between the predicted and actual results.

What error? I wasn't instructed to run or look at anything, so how can I know there's an error?

> Next, lets create an asset which runs inference on the model, more frequently than the model is re-trained to predict the results.

I'm finding this hard to follow - I think it's because of how it's structured. Is this accurate?:

> Next, to predict the results, we'll create another asset that runs inference on the model more frequently than the model is re-trained.

File: docs/content/guides/dagster/ml-ops.mdx [hunk 21]
```diff
@@ -0,0 +1,221 @@
...
+    df = pd.DataFrame(results)
+    if len(df) > 0:
+        df = df[df.type == "story"]
+        df = df[~df.title.isna()]
+    inference_x = df.title
+    inference_x = Tfidf_Vectorizer.transform(inference_x)
+    return xgboost.predict(inference_x)
+```
+
+Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history and trigger retraining.
```
Comment (erinkcochran87) [inline docs/content/guides/dagster/ml-ops.mdx:191]:
```suggestion
Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history, and trigger retraining.
```

🔧 Oxford comma pls

File: docs/content/guides/dagster/ml-ops.mdx [hunk 22]
```diff
@@ -0,0 +1,221 @@
...
+
+Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.
```
Comment (erinkcochran87) [inline docs/content/guides/dagster/ml-ops.mdx:193]: 🤔 What's the user supposed to do with this information? Use W&B in their own models? Currently, it feels a bit tacked on - it could use some additional info.

**Review by erinkcochran87 — COMMENTED** [review]
Back to you! I agree with Sandy's general comments on grammar/polish. Wanna try running this through Grammarly before your next pass?

File: docs/content/guides/dagster/ml-ops.mdx (hunk 20 above)
Comment (odette-elementl) [reply docs/content/guides/dagster/ml-ops.mdx:165]: updated- do we know how to put R^2 in nicely?

File: docs/content/guides/dagster/ml-ops.mdx (hunk 22 above)
Comment (odette-elementl) [reply docs/content/guides/dagster/ml-ops.mdx:195]: added more of a section with some other recommendations

File: docs/content/guides/dagster/ml-ops.mdx [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+
+First, we will create an asset that retrieves the most recent Hacker News records.
+
+```python file=/guides/dagster/ml_ops/ml_ops.py startafter=data_ingestion_start endbefore=data_ingestion_end
+import requests
+from dagster import asset, FreshnessPolicy
+import pandas as pd
+
+
+@asset(freshness_policy=FreshnessPolicy(maximum_lag_minutes=10))
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:62]: I think we should probably actually leave out the `FreshnessPolicy`, at least for now. Because we're not telling people how to schedule their model, so its presence is likely to be confusing.

I also think they might not necessarily want to update their training data every 10 minutes.

File: docs/content/guides/dagster/ml-ops.mdx [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+
+Depending on what the objective of your ML model is, you can use this data to set alerts, save model performance history, and trigger retraining.
+
+---
+
+## Where to go from here
+
+- Dagster is flexible so there could be many 'right' ways to implement your machine learning pipeline and different avenues to explore
+- Dagster intergrates with [MLflow](/\_apidocs/libraries/dagster-mlflow) that can be used to keep track of your models
+- Dagster integrates with [Weights & Biases](/\_apidocs/libraries/dagster-wandb) and an [example](https://github.com/dagster-io/dagster/tree/master/examples/with_wandb) which demonstrates how to use W\&B's artifacts with Dagster.
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:211]: A backslash is showing up when this is rendered

File: docs/content/guides/dagster/ml-ops.mdx [hunk 25]
```diff
@@ -0,0 +1,211 @@
+---
+title: Building machine learning pipelines In Dagster | Dagster Docs
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:2]: "In" should not be capitalized

File: docs/content/guides/dagster/ml-ops.mdx [hunk 26]
```diff
@@ -0,0 +1,211 @@
...
+description: This guide illustrates how to use Dagster to operationalize your machine learning pipeline
+---
+
+# Building machine learning pipelines with Dagster
+
+In this guide, we’ll walk you through how to take your machine learning models and deploy and maintain them in production using Dagster, reliably and efficiently.
+
+We will work through building a machine lerning pipeline, including using assets for different elements, how to automate model training, and monitoring your model's drift.
```
Comment (sryza) [inline docs/content/guides/dagster/ml-ops.mdx:10]:
```suggestion
We will work through building a machine learning pipeline, including using assets for different elements, how to automate model training, and monitoring your model's drift.
```

**Review by sryza — COMMENTED** [review]
Just a few more comments and then this looks good to go!

Let me know if you need help getting this to pass buildkite

File: docs/content/guides/dagster/ml-ops.mdx (hunk 23 above)
Comment (odette-elementl) [reply docs/content/guides/dagster/ml-ops.mdx:62]: ok removed from this and inference

File: docs/content/guides/dagster/ml-ops.mdx [hunk 27]
```diff
@@ -0,0 +1,211 @@
...
+
+For this guide, we will be using the Hacker News data demoed in the [tutorial](/tutorial).
+
+The machine learning model we will walk through takes the Hacker News stories and uses the titles to predict the number of comments that a story will generate. This will be a supervised model since we have the number of comments for all the previous stories.
+
+The assets graph will look like this at the end of this guide:
+
+<Image
+alt="alt"
+src="/images/guides/ml-ops/ml_asset_dag.png"
```
[the remark this answers is not in the document]
Comment (odette-elementl) [reply docs/content/guides/dagster/ml-ops.mdx]: that should be better!

**Comment by sryza** [comment]
@smackesey - do you have an answer to this question from @odette-elementl?

> @sryza pyright is saying to add xgboost to pyright/master/requirements.txt, It looks like we have everything in requirements-pinned.txt, is it cool to edit that and add it?

File: docs/content/guides/dagster/ml-ops.mdx (hunk 24 above)
Comment (erinkcochran87) [reply docs/content/guides/dagster/ml-ops.mdx]: This is still an issue for me as well

**Review by erinkcochran87 — COMMENTED** [review]
Just a few small things - this is so close and looking great!

**Comment by smackesey** [comment]
@odette-elementl You shouldn't directly edit `requirements-pinned.txt`, instead run `make rebuild_pyright_pins`. That will rebuild the venv with your dep (assuming it is specified in some setup.py) and update `requirements-pinned.txt`.

**Comment by odette-elementl** [comment]
> @odette-elementl You shouldn't directly edit `requirements-pinned.txt`, instead run `make rebuild_pyright_pins`. That will rebuild the venv with your dep (assuming it is specified in some setup.py) and update `requirements-pinned.txt`.

If it wasnt in a setup.py -> do i need to create one? this was just importing a package to use on an example

**Comment by smackesey** [comment]
>If it wasnt in a setup.py -> do i need to create one? this was just importing a package to use on an example

Looking at the diff, you're using `xgboost` in `docs_snippets`, so you should add it to the `extra_require[full]` section of `docs_snippets/setup.py`: https://github.com/dagster-io/dagster/blob/master/examples/docs_snippets/setup.py#L37-L58

**Comment by odette-elementl** [comment]
conflicts so created PR 13100
